pkg/tfbridge/defaults/tokens.go

```diff
@@ -0,0 +1,928 @@
+)
+
+const (
+	defaultMinimumModuleSize    = 5
+	defaultMinimumSubmoduleSize = 5
+)
+
+// A function that joins a module and name into a pulumi type token.
+//
+// For example:
+//
+//	func(module, name string) (string, error) {
+//	 return fmt.Sprintf("pkgName:%s:%s", module, name), nil
+//	}
+type MakeToken func(module, name string) (string, error)
+
+// A strategy that assigns all tokens to the same module.
+//
+// For example:
+//
+//	rStrat, dStrat := TokensSingleModule("pkgName_", "index", finalize)
+//
+// The above example would transform "pkgName_foo" into "pkgName:index:Foo".
+func TokensSingleModule(
+	tfPackagePrefix, moduleName string, finalize MakeToken,
+) Strategy {
+	return TokensKnownModules(tfPackagePrefix, moduleName, nil, finalize)
+}
+
+func tokensKnownModules[T b.ResourceInfo | b.DataSourceInfo](
+	prefix, defaultModule string, modules []string,
+	new func(string, string) (*T, error),
+	moduleTransform func(string) string,
+) ElementStrategy[T] {
+	return func(tfToken string) (*T, error) {
+		tk := strings.TrimPrefix(tfToken, prefix)
+		if len(tk) == len(tfToken) {
+			return nil, fmt.Errorf("token '%s' missing package prefix '%s'", tfToken, prefix)
+		}
+		mod := defaultModule
+		for _, m := range modules {
+			if strings.HasPrefix(tk, m) {
+				mod = m
+				break
+			}
+		}
+		if mod == "" {
+			return nil, fmt.Errorf("could not find a module that prefixes '%s' in '%#v'", tk, modules)
+		}
+		return new(moduleTransform(mod), upperCamelCase(strings.TrimPrefix(tk, mod)))
+	}
+}
+
+// A strategy for assigning tokens to a hand generated set of modules.
+//
+// If defaultModule is "", then the returned strategies will error on not encountering a matching module.
+func TokensKnownModules(
+	tfPackagePrefix, defaultModule string, modules []string, finalize MakeToken,
+) Strategy {
+	// NOTE: We could turn this from a sort + linear lookup into a radix tree to recover
+	// O(log(n)) performance (current is O(n*m)) where n = number of modules and m =
+	// number of mappings.
+	sort.Sort(sort.Reverse(sort.StringSlice(modules)))
+
+	return Strategy{
+		Resource: tokensKnownModules(tfPackagePrefix, defaultModule, modules,
+			func(mod, tk string) (*b.ResourceInfo, error) {
+				tk, err := finalize(mod, tk)
+				if err != nil {
+					return nil, err
+				}
+				return &b.ResourceInfo{Tok: tokens.Type(tk)}, nil
+			}, camelCase),
+		DataSource: tokensKnownModules(tfPackagePrefix, defaultModule, modules,
+			func(mod, tk string) (*b.DataSourceInfo, error) {
+				tk, err := finalize(mod, "get"+tk)
+				if err != nil {
+					return nil, err
+				}
+				return &b.DataSourceInfo{Tok: tokens.ModuleMember(tk)}, nil
+			}, camelCase),
+	}
+}
+
+// A strategy for assigning tokens to a hand generated set of modules with an arbitrary
+// mapping from TF modules to Pulumi modules.
+//
+// If defaultModule is "", then the returned strategies will error on not encountering a matching module.
+func TokensMappedModules(
```

Comment:
Again, quick example would be great.

Just curious - is generalizing `defaultModule string, modules map[string]string` to `func(string)string` not feasible for fundamental reasons? I thought perhaps if the system has access to the entire set of TF tokens upfront it could do this:

```
m := map[string]string{}
for _, t := range allTokens {
   m[t] = f(t)
}
return m
```

To convert a functional assignment strategy to a map.

Reply:
It can, but that is just implementing a new strategy. Having a pre-computed map allows us to implement more complicated logic that otherwise the user would need to implement.

For example:

```go
TokensMappedModules("index", map[string]string{
    "mod": "mod",
    "mod_nest": "nested",
}, MakeStandardToken("example"))
```

Consider the token `"mod_nest_one"`. `TokensMappedModules` handles the logic to determine that `mod_nest_one` maps to `example:nested/one:One` instead of `example:mod/nestOne:NestOne`, even though without `mod_nest` as a prefix, it would map to `example:mod/nestOne:NestOne`.

---

We could expose a helper function that makes it easier to implement your own strategy but it would not replace `TokensMappedModules`.